### PR TITLE
Merged iframe and web workers together into a new package

### LINKED ISSUE #225: Web Worker execution environment
We should get the web worker execution environment working with Flask. The web worker environment might be required for manifest v3 compatibility.

### PR DESCRIPTION
**Pull request by ritave**
After the new package is uploaded to npm we can remove workers package

fixes #225

### CODE REVIEW

**Comment by rekmarks** [comment]
@ritave did you test the Web Worker environment on Flask in place of the iframe?

**Comment by ritave** [comment]
@rekmarks 

> @ritave did you test the Web Worker environment on Flask in place of the iframe? That's not a requirement for this PR, but we'd want to confirm that it boots before closing #225.

Yeah I got it working. Ready for review

**Comment by shanejonas** [comment]
the bundle seems a lot bigger:

```WARNING in entrypoint size limit: The following entrypoint(s) combined asset size exceeds the recommended limit (244 KiB). This can impact web performance.
Entrypoints:
  iframe (346 KiB)
      iframe.bundle.js
  webworker (346 KiB)
      webworker.bundle.js
```

vs [70k](https://metamask.github.io/iframe-execution-environment/0.3.1/bundle.js)

is it the `resolve` for stream?

**Comment by shanejonas** [comment]
ah ok it looks like its about the same size after gzipping:

image

File: packages/execution-environments/src/openrpc.json [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+              "value": "TestSnap"
+            },
+            {
+              "name": "origin",
+              "value": "foo.com"
+            },
+            {
+              "name": "request",
+              "value": {
+                "method": "hello",
```
Comment (shanejonas) [inline packages/execution-environments/src/openrpc.json:133]:
```suggestion
                "jsonrpc": "2.0",
                "method": "hello",
```

File: packages/execution-environments/webpack.config.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    mode: argv.mode,
+    entry: {
+      iframe: './src/iframe/index.ts',
+      webworker: './src/web-workers/index.ts',
+    },
+    output: {
+      filename: '[name].bundle.js',
+      path: DIST,
+    },
+    plugins: [new NodePolyfillPlugin()],
```
Comment (rekmarks) [inline packages/execution-environments/webpack.config.js:27]: We should ensure that SES lockdown is copied purely as in [the orignal implementation](https://github.com/MetaMask/iframe-execution-environment/blob/main/webpack.config.js#L35-L42), and added to `index.html` via its own `<script>` tag, which should also be the first script to execute.

**Review by rekmarks — CHANGES_REQUESTED** [review]

File: packages/execution-environments/src/common/lockdown/lockdown-more.ts [hunk 3]
```diff
@@ -0,0 +1,90 @@
+/* eslint-disable-next-line import/no-unassigned-import */
+import 'ses';
```
Comment (rekmarks) [inline packages/execution-environments/src/common/lockdown/lockdown-more.ts:2]: Per my other comment, [`ses` should not be imported in this manner](https://github.com/MetaMask/snaps-skunkworks/pull/231/files#r807333562).

File: packages/execution-environments/LICENSE [hunk 4]
```diff
@@ -0,0 +1,18 @@
+Copyright ConsenSys Software Inc. 2021. All rights reserved.
```
Comment (rekmarks) [inline packages/execution-environments/LICENSE:1]: nit:
```suggestion
Copyright ConsenSys Software Inc. 2022. All rights reserved.
```

File: packages/execution-environments/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# @metamask/execution-environments
+
+Snap sandbox environments for executing SES javascript
```
Comment (rekmarks) [inline packages/execution-environments/README.md:3]:
```suggestion
Execution environments for [MetaMask Snaps](https://docs.metamask.io/guide/snaps.html), sandboxed using [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses).
```

File: packages/execution-environments/package.json [hunk 6]
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "@metamask/execution-environments",
+  "version": "0.1.0",
```
Comment (rekmarks) [inline packages/execution-environments/package.json:3]: non-blocking: If we start the package at version `0.0.0`, it tends to work better with our release automation, although it probably doesn't matter for monorepos as long as the version is less than or equal to the highest version in the monorepo.

```suggestion
  "version": "0.0.0",
```

File: packages/execution-environments/package.json [hunk 7]
```diff
@@ -0,0 +1,67 @@
+{
+  "name": "@metamask/execution-environments",
```
Comment (rekmarks) [inline packages/execution-environments/package.json:2]: I think we have an opportunity to improve on our previously established naming conventions here:

```suggestion
  "name": "@metamask/snap-executors",
```

File: packages/execution-environments/src/common/BaseControllers.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+  origin: string,
+  request: JsonRpcRequest,
+) => Promise<unknown>;
+
+const fallbackError = {
+  code: errorCodes.rpc.internal,
+  message: 'Execution Environment Error',
+};
+
+export class BaseController {
```
Comment (rekmarks) [inline packages/execution-environments/src/common/BaseControllers.ts:28]: In line with the suggested renaming of this package:

```suggestion
export class BaseSnapExecutor {
```

File: packages/execution-environments/src/common/lockdown/lockdown.ts [hunk 9]
```diff
@@ -0,0 +1,18 @@
+/* eslint-disable-next-line import/no-unassigned-import */
+import 'ses';
```
Comment (rekmarks) [inline packages/execution-environments/src/common/lockdown/lockdown.ts:2]: We should get rid of this `ses` import, too. For future reference, SES only needs to be imported once, and it mutates the global environment (realm) automagically when it's imported.

File: packages/execution-environments/src/iframe/IFrameController.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import ObjectMultiplex from '@metamask/object-multiplex';
+import { WindowPostMessageStream } from '@metamask/post-message-stream';
+import pump from 'pump';
+import { BaseController } from '../common/BaseControllers';
+import { SNAP_STREAM_NAMES } from '../common/enums';
+
+export class IFrameController extends BaseController {
```
Comment (rekmarks) [inline packages/execution-environments/src/iframe/IFrameController.ts:7]: In line with the suggested renaming:

```suggestion
export class IFrameSnapExecutor extends BaseSnapExecutor {
```

File: packages/execution-environments/src/web-workers/WebWorkerController.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import ObjectMultiplex from '@metamask/object-multiplex';
+import { WorkerPostMessageStream } from '@metamask/post-message-stream';
+import pump from 'pump';
+import { BaseController } from '../common/BaseControllers';
+import { SNAP_STREAM_NAMES } from '../common/enums';
+
+export class WebWorkerController extends BaseController {
```
Comment (rekmarks) [inline packages/execution-environments/src/web-workers/WebWorkerController.ts:7]: In line with the suggested renaming:

```suggestion
export class WebWorkerSnapExecutor extends BaseSnapExecutor {
```

**Review by rekmarks — CHANGES_REQUESTED** [review]
Nicely done! Aside from the `ses` import / initialization stuff, my main feedback regards renaming ~this package and its~ the classes in this package to "snap executors" instead of "controllers".

**Comment by rekmarks** [comment]
Previously, I suggested that we rename this package to `snap-executors` in my most recent review. On second thought, I think that's a bad idea, and the package itself should refer to "execution environments". I still think we should use names other than "controller" to refer to the classes that manage the execution of snaps inside of our execution environments, however.

File: packages/execution-environments/src/web-workers/index.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+//------------------------
+// Notice that, compared to the rest of the package, we're including ses directly here, instead of just referencing types.
+// This is because IFrame has ses included as a seperate <script> tag, while WebWorkers is just one file.
+// We ensure ses is the first thing to be executed in the index file of WebWorker code
+/* eslint-disable-next-line import/no-unassigned-import */
+import 'ses';
```
Comment (shanejonas) [inline packages/execution-environments/src/web-workers/index.ts:6]: we shouldnt include this in the bundle see https://github.com/MetaMask/snaps-skunkworks/pull/231/files#r807333562

File: packages/execution-environments/src/common/BaseSnapExecutor.ts [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+    };
+
+    try {
+      const compartment = new Compartment({
+        ...endowments,
+        window: { ...endowments },
+      });
+      compartment.evaluate(sourceCode);
+
+      self?.addEventListener(
```
Comment (rekmarks) [inline packages/execution-environments/src/common/BaseSnapExecutor.ts:208]: follow-up: We should fix the types such that `self` is always defined, if possible.

**Review by rekmarks — APPROVED** [review]
Excellent. LGTM!

**Review by rekmarks — APPROVED** [review]

**Pull request merged**